metadata/src/utils/validation.rs

```diff
@@ -215,6 +240,57 @@ fn get_extrinsic_hash(
     bytes
 }
 
+/// Obtain the hash representation of the `frame_metadata::v15::OuterEnums`.
+fn get_outer_enums_hash(
+    registry: &PortableRegistry,
+    enums: &OuterEnumsMetadata,
+    only_these_variants: Option<&[&str]>,
+) -> [u8; HASH_LEN] {
+    let mut visited_ids = HashSet::<u32>::new();
+
+    /// Hash the provided enum type.
+    fn get_enum_hash(
+        registry: &PortableRegistry,
+        id: u32,
+        only_these_variants: Option<&[&str]>,
+        visited_ids: &mut HashSet<u32>,
+    ) -> [u8; HASH_LEN] {
+        let ty = registry
+            .types
+            .get(id as usize)
+            .expect("Metadata should contain enum type in registry");
+
+        if let TypeDef::Variant(variant) = &ty.ty.type_def {
+            get_type_def_variant_hash(registry, variant, only_these_variants, visited_ids)
```

Comment:
I'm pondering the use of `visited_ids` a bit now; eg is it right to pass the same `visited_ids` set into the call for each enum? What's the effect of doing so versus not doing so.

Doing so means that hashing things that share types might be faster; any type that's been seen will hash to some fixed "recursive" hash. The downside though is that the individual hashes become "weaker"; a bunch of more primitive types all are assigned the "recursive" hash when hashing the second and third enum. This makes them all effectively interchangeable; we'll no longer catch changes to them.

The point of `visited_ids`  is only to prevent infinite loops, but given the above, let's pass an empty `visited_ids` set into each new top level call; so here `get_enum_hash` could create an empty Set each time for instance so that the types are entirely independently hashed.